src/cli/NativeMsgIface.cpp

```diff
@@ -0,0 +1,430 @@
+    });
+#else
+    return QObject::connect(sender, signal, context, slot, Qt::SingleShotConnection);
+#endif
+}
+
+// Little helper to print QSet<QString> and QList<QString> without the need to
+// convert them into a QStringList.
+template <typename T>
+QString join(QString glue, T const &list) {
+    QString out;
+    for (auto &&item : list) {
+        if (!out.isEmpty())
+            out += glue;
+        out += item;
+    }
+    return out;
+};
+
+}
+
+NativeMsgIface::NativeMsgIface(QObject * parent) :
+      QObject(parent)
+      , network_(this)
+      , settings_(this)
+      , models_(this, &settings_)
+      , operations_(0)
+      , die_(false)
+    {
+    qRegisterMetaType<std::shared_ptr<std::vector<char>>>("std::shared_ptr<std::vector<char>>");
+    
+    // Disable synchronisation with C style streams. That should make IO faster
+    std::ios_base::sync_with_stdio(false);
+
+    // Init the marian translation service:
+    marian::bergamot::AsyncService::Config serviceConfig;
+    serviceConfig.numWorkers = settings_.marianSettings().cpu_threads;
+    serviceConfig.cacheSize = settings_.marianSettings().translation_cache ? kTranslationCacheSize : 0;
+    service_ = std::make_shared<marian::bergamot::AsyncService>(serviceConfig);
+
+    // Pick up on network errors
+    // TODO: not sure if every request always returns on 1 error. fetchRemoteModels for example
+    // will cut short if a request is already in progress. If that one fails, the second request
+    // that got cut short will have no response, and stay pending.
+    connect(&network_, &Network::error, this, [&](QString err, QVariant data) {
+        if (data.canConvert<Request>())
+            writeError(data.value<Request>(), std::move(err));
+    });
+
+    connect(&network_, &Network::downloadComplete, this, [this](QFile *file, QString filename, QVariant data) {
+        models_.writeModel(file, filename);
+
+        if (data.canConvert<DownloadRequest>()) {
+            DownloadRequest request = data.value<DownloadRequest>();
+            writeResponse(request, QJsonObject{{"modelID", request.modelID}});
+        }
+    });
+
+    connect(this, &NativeMsgIface::emitJson, this, &NativeMsgIface::processJson);
+}
+
+void NativeMsgIface::run() {
+    iothread_ = std::thread([this](){
+        for (;;) {
+            // First part of the message: Find how long the input is. If that
+            // read fails, we're probably at EOF.
+            char len[4];
+            if (!std::cin.read(len, 4))
+                break;
+
+            int ilen = *reinterpret_cast<unsigned int *>(len);
+            if (ilen >= kMaxInputLength || ilen < 2) { // >= 2 because JSON is at least "{}"
+                std::cerr << "Invalid message size. Shutting down." << std::endl;
+                break;
+            }
+
+            //  Read in the message into Json
+            std::shared_ptr<std::vector<char>> input(std::make_shared<std::vector<char>>(ilen));
```

Comment:
What's the benefit of using `shared_ptr` here, if we are already using the heavyweight vector? I think `std::array` would be a better choice, since the input is constant size?

Reply:
The shared_ptr is to avoid copies as the input goes through Qt's signal/slot interface. The vector because macOS 10.15 ships with a compiler that doesn't like smart pointers and C-style arrays for some reason. `std::array` won't work here because the size of that thing is a template variable, not a runtime variable.

But I just replaced all of it with a QByteArray, which was the type it will be converted to anyway! And QByteArray also does all the copy-on-write Qt smart copy stuff in the background, so no need for a shared_ptr :)